### PR TITLE
[8.0] Backport ResolveSE from LHCbDIRAC

### PR DESCRIPTION
**Pull request by chaen**
This is quite a bulky PR but a lot of it is docs moving around, so do not be scared

The most interesting things are:
* d4cd3399f8c69a270af4831e6f94f99dce0f9d04:  the introduction of a test utility to have temporary local DIRAC CFG 
* d13941d4e212ed0b169575e2abdc3815f7c1bfce: backport getDestinationSEList from LHCbDIRAC (already used by ILC). This basically allows to overwrite StorageElementGroup definition at a site or Country level. This is useful for job ouput upload.
* 87dae30ca79dc87e9f724fbdf3fad7979634ba00 JobWrapper uses getDestinationSEList, which should be totally transparent for VO without AssociatedSE configuration (actually, the order of the returned SE may change, but not the actual list)

BEGINRELEASENOTES

*Tests
NEW: introduce generateDIRACConfig utility to generate DIRAC config during tests

*Docs:
CHANGE: move site documentation from ConfigRef to Resources and dirac.cfg
NEW: document Countries

*DMS
NEW: backport getDestinationSEList from LHCbDIRAC

*WMS
CHANGE: JobWrapper uses getDestinationSEList rather that resolveSEGroup

ENDRELEASENOTES

### CODE REVIEW

File: docs/source/DMSOpsDocs.rst [hunk 1]
```diff
@@ -0,0 +1,7 @@
+Describe how input data is treated
+
+describe how outptu data is treated
```
Comment (phicharp) [inline docs/source/DMSOpsDocs.rst:3]: Typo

**Comment by chaen** [comment]
@andresailer it's not completely finished yet, but you may already want to have a look :-)

File: dirac.cfg [hunk 2]
```diff
@@ -852,6 +852,162 @@ Resources
     # This option is used in the Job Wrapper and, if set, requires the RequestManagementSystem to be installed
     Tier1-Failover = CERN-FAILOVER,CNAF-FAILOVER
   }
+  # Definition of the sites
+  # See file:///home/chaen/dirac/DIRAC/docs/build/html/AdministratorGuide/Resources/site.html
```
Comment (fstagni) [inline dirac.cfg]: I guess this is your note.

File: dirac.cfg [hunk 3]
```diff
@@ -852,6 +852,162 @@ Resources
...
+        SE = CERN-RAW, CERN-RDST, CERN-USER # (Optional) SEs Local to the site
+
+        # Overwrite definities of StorageElement (discouraged)
+        # or StorageElementGroups for that Site
+        AssociatedSEs
+        {
+          # Tier1-Failover is now only CERN-FAILOVER when running a Job at CERN
+          Tier1-Failover = CERN-FAILOVER
+        }
+        CE = ce503.cern.ch, ce504.cern.ch # List of CEs for th esite (updated by BDII2CSAgent)
```
Comment (fstagni) [inline dirac.cfg]:
```suggestion
        CE = ce503.cern.ch, ce504.cern.ch # List of CEs for the site (updated by hand, or by BDII2CSAgent)
```

File: dirac.cfg [hunk 4]
```diff
@@ -852,6 +852,162 @@ Resources
...
+        Name = CERN-PROD # (Optional) Name of the site from the admin, e.g in GOCDB
+        Coordinates = 06.0458:46.2325 # (Optional) Geographical coordinates
+        Mail = grid-cern-prod-admins@cern.ch # (Optional) Geographical coordinates
```
Comment (fstagni) [inline dirac.cfg]:
```suggestion
        Mail = grid-cern-prod-admins@cern.ch # (Optional) site admin mail
```

File: dirac.cfg [hunk 5]
```diff
@@ -852,6 +852,162 @@ Resources
...
+        MoUTierLevel = 0 # (Optional) Tier level
+        Description = CERN European Organization for Nuclear Research # (Optional) ...
+        # Subsection to describe each CE available
```
Comment (fstagni) [inline dirac.cfg:879]:
```suggestion
        # Subsection to describe each Computing Element available
```

File: dirac.cfg [hunk 6]
```diff
@@ -852,6 +852,162 @@ Resources
...
+            # CE architecture
+            architecture = x86_64
+
+            # CE operating system in a DIRAC format
+            OS = ScientificCERNSLC_Carbon_6.4
+
+            # Boolean attributes than indicates if the site accept pilots
+            Pilot = False
+
+            # Type of CE, can take values as LCG or CREAM
```
Comment (fstagni) [inline dirac.cfg]:
```suggestion
            # Type of CE, can take any CE type DIRAC recognizes (:ref:`documentation of the CEs <CE>`)
```

File: dirac.cfg [hunk 7]
```diff
@@ -852,6 +852,162 @@ Resources
...
+            # Type of 'Inner' CE, normally empty. Default = "InProcess".
+            # Possibilities: potentially all CE types, but in practice
+            # the most valid would be: InProcess, Sudo, Singularity, Pool.
+            # Pool CE in turn uses InProcess (Default)
+            # or Sudo or Singularity. To specify, use Pool/ce_type.
+            # This option can also go at the Queue level.
+            LocalCEType = Pool
+
+            # If the CE is a cream CE the mode of submission
+            SubmissionMode = Direct
```
Comment (fstagni) [inline dirac.cfg:907]: This is probably not necessary anymore (to be checked, probably to be referenced in new issue)

File: dirac.cfg [hunk 8]
```diff
@@ -852,6 +852,162 @@ Resources
...
+            wnTmpDir = .
+
+            # CE architecture
+            architecture = x86_64
+
+            # CE operating system in a DIRAC format
+            OS = ScientificCERNSLC_Carbon_6.4
+
+            # Boolean attributes than indicates if the site accept pilots
+            Pilot = False
```
Comment (fstagni) [inline dirac.cfg]: These might not be necessary anymore (to be checked, probably to be referenced in new issue)

File: dirac.cfg [hunk 9]
```diff
@@ -852,6 +852,162 @@ Resources
...
+            # the most valid would be: InProcess, Sudo, Singularity, Pool.
+            # Pool CE in turn uses InProcess (Default)
+            # or Sudo or Singularity. To specify, use Pool/ce_type.
+            # This option can also go at the Queue level.
+            LocalCEType = Pool
+
+            # If the CE is a cream CE the mode of submission
+            SubmissionMode = Direct
+
+            # Number of available processors on worker nodes
```
Comment (fstagni) [inline dirac.cfg]:
```suggestion
            # (Optional) max number of processors that DIRAC pilots are allowed to exploit. Implicit default = 1
```

File: dirac.cfg [hunk 10]
```diff
@@ -852,6 +852,162 @@ Resources
...
+            # CE operating system in a DIRAC format
+            OS = ScientificCERNSLC_Carbon_6.4
+
+            # Boolean attributes than indicates if the site accept pilots
+            Pilot = False
+
+            # Type of CE, can take values as LCG or CREAM
+            CEType = HTCondorCE
+
+            # Type of 'Inner' CE, normally empty. Default = "InProcess".
```
Comment (fstagni) [inline dirac.cfg]:
```suggestion
            # (Optional) Type of 'Inner' CE, normally empty. Default = "InProcess".
```

File: dirac.cfg [hunk 11]
```diff
@@ -852,6 +852,162 @@ Resources
...
+            # This option can also go at the Queue level.
+            LocalCEType = Pool
+
+            # If the CE is a cream CE the mode of submission
+            SubmissionMode = Direct
+
+            # Number of available processors on worker nodes
+            NumberOfProcessors = 12
+
+            # Number of available worker nodes per allocation.
```
Comment (fstagni) [inline dirac.cfg]:
```suggestion
            # (Optional) Number of available worker nodes per allocation. Implicit default = 1
```

File: dirac.cfg [hunk 12]
```diff
@@ -852,6 +852,162 @@ Resources
...
+            # Values can be a number (e.g. 2 nodes) or a range of values
+            # (e.g. from 2 to 4 nodes) which lets the choice to the batch
+            # system.
+            NumberOfNodes = 2
+            # NumberOfNodes = 2-4
+
+            # CE allows *whole node* jobs
```
Comment (fstagni) [inline dirac.cfg]:
```suggestion
            # (Optional) CE allows *whole node* jobs. Implicit default = False
```

File: dirac.cfg [hunk 13]
```diff
@@ -852,6 +852,162 @@ Resources
...
+            WholeNode = True
+
+            # List of tags specific for the CE
```
Comment (fstagni) [inline dirac.cfg]:
```suggestion
            # (Optional) List of tags specific for the CE
```

File: dirac.cfg [hunk 14]
```diff
@@ -852,6 +852,162 @@ Resources
...
+            Tag = GPU, 96RAM
+
+            # List of required tags that a job to be eligible must have
```
Comment (fstagni) [inline dirac.cfg]:
```suggestion
            # (Optional) List of required tags that a job to be eligible must have
```

File: docs/source/AdministratorGuide/Configuration/ConfReference/Resources/FileCatalogs/index.rst [hunk 15]
```diff
@@ -1,16 +0,0 @@
-Resources / FileCatalogs - Subsections
-======================================
-
-This subsection include the definition of the File Catalogs to be used in the installation. In case there is more than one File Catalog defined in this section, the first one in the section will be used as default by the ReplicaManager client.
-
```
Comment (fstagni) [inline docs/source/AdministratorGuide/Configuration/ConfReference/Resources/FileCatalogs/index.rst:5]: Are the below entries still valid? You have not added them to dirac.cfg in this PR.

File: docs/source/AdministratorGuide/Resources/site.rst [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+=====
+
+
+Site Names
+----------
+
+Sites have names resulting from concatenation of:
+
+- Domain: Grid site name, expressed in uppercase, for example: LCG, EELA
+- Site: Institution acronym in uppercase, for example: CPPM
```
Comment (fstagni) [inline docs/source/AdministratorGuide/Resources/site.rst]: not necessarily uppercase

File: src/DIRAC/DataManagementSystem/Utilities/ResolveSE.py [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+    if result["OK"]:
+        sLog.info("Found concrete SE", str(outputSE))
+        return [outputSE]
+
+    # Get local SEs
+    localSEs = getSEsForSite(site)
+    if not localSEs["OK"]:
+        raise RuntimeError(localSEs["Message"])
+    localSEs = localSEs["Value"]
+    sLog.verbose("Local SE list is:", str(localSEs))
```
Comment (fstagni) [inline src/DIRAC/DataManagementSystem/Utilities/ResolveSE.py]:
```suggestion
    sLog.verbose("Local SE list is: ", ", ".join(localSEs))
```

File: src/DIRAC/DataManagementSystem/Utilities/test/Test_resolveSE.py [hunk 18]
```diff
@@ -0,0 +1,307 @@
...
+    """Should return the LogSE as is, full stop"""
+    # Asking for a specific SE should return that SE, no matter what
+    for site in ("LCG.CERN.cern", "LCG.CNAF.it", "LCG.IN2P3.fr", "LCG.GRIDKA.de", "LCG.NCBJ.pl", "AnySite"):
+        assert getDestinationSEList("LogSE", site) == ["LogSE"]
+        assert getDestinationSEList("LogSE", site, outputmode="Local") == ["LogSE"]
+
+
+def test_emptyInput():
+    """Should return empty list"""
+    # Asking for a specific SE should return that SE, no matter what
```
Comment (fstagni) [inline src/DIRAC/DataManagementSystem/Utilities/test/Test_resolveSE.py]:
```suggestion
    # Asking for no SE should return an empty list, no matter what
```

File: src/DIRAC/WorkloadManagementSystem/JobWrapper/JobWrapper.py [hunk 19]
```diff
@@ -43,6 +43,7 @@
 from DIRAC.DataManagementSystem.Client.DataManager import DataManager
 from DIRAC.DataManagementSystem.Client.FailoverTransfer import FailoverTransfer
 from DIRAC.DataManagementSystem.Utilities.DMSHelpers import resolveSEGroup
```
Comment (fstagni) [inline src/DIRAC/WorkloadManagementSystem/JobWrapper/JobWrapper.py]: Then this is not used, at least in this module. It is still used in other places, though, should those be consider for an update?

**Comment by chaen** [comment]
@fstagni regarding all the typo and mistakes in the WMS docs, I Wanted to ask you to have a look, because I am pretty sure that a lot of things are not valid anymore. I propose that you do all the relevant changes directly on this PR, you should have push permissions

File: src/DIRAC/WorkloadManagementSystem/JobWrapper/JobWrapper.py (hunk 19 above)
Comment (chaen) [reply src/DIRAC/WorkloadManagementSystem/JobWrapper/JobWrapper.py]: correct

Comment (chaen) [reply src/DIRAC/WorkloadManagementSystem/JobWrapper/JobWrapper.py]: For the other places, not necessarily, it only makes sense for job upload

File: docs/source/AdministratorGuide/Configuration/ConfReference/Resources/FileCatalogs/index.rst (hunk 15 above)
Comment (chaen) [reply docs/source/AdministratorGuide/Configuration/ConfReference/Resources/FileCatalogs/index.rst:5]: the necessary was already ported long ago

File: dirac.cfg (hunk 2 above)
Comment (chaen) [reply dirac.cfg]: correct :man_facepalming:

File: docs/source/AdministratorGuide/Resources/site.rst [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+- Country: country where the site is located, expressed in lowercase, for example fr
```
Comment (andresailer) [inline docs/source/AdministratorGuide/Resources/site.rst:14]: if the `.cern` is a valid sitename, then this is the top-level-domain and not country code?

File: src/DIRAC/DataManagementSystem/Utilities/ResolveSE.py [hunk 21]
```diff
@@ -0,0 +1,108 @@
...
+    shuffle(seList)
+    # localSEs are put first in the list
+    return sorted(seList, key=lambda x: x not in localSEs)
+
+
+def getDestinationSEList(outputSE, site, outputmode="Any"):
+    """Evaluate the output SE list from a workflow and return the concrete list
+    of SEs to upload output data. The resolution order goes as follow:
+
+        * outputSE as a normal StorageElement
```
Comment (andresailer) [inline src/DIRAC/DataManagementSystem/Utilities/ResolveSE.py]: indentation not needed for bullet list

File: src/DIRAC/DataManagementSystem/Utilities/ResolveSE.py [hunk 22]
```diff
@@ -0,0 +1,108 @@
...
+        * outputSE as an alias of one SE defined in the ``site`` AssociatedSEs (return local first)
+        * outputSE as an alias of multiple SE (local SE should come first)
+        * outputSE as a StorageElementGroup
+
+    Moreover, if
```
Comment (andresailer) [inline src/DIRAC/DataManagementSystem/Utilities/ResolveSE.py]: Merge with next line, and drop one of the ifs

File: src/DIRAC/DataManagementSystem/Utilities/ResolveSE.py [hunk 23]
```diff
@@ -0,0 +1,108 @@
...
+    If output mode is `Local`:
+
+        * return ONLY local SE within the SEGroup if they exist (i.e. in the ``<site>/SE>`` config)
```
Comment (andresailer) [inline src/DIRAC/DataManagementSystem/Utilities/ResolveSE.py]: indentation not needed

File: src/DIRAC/DataManagementSystem/Utilities/ResolveSE.py [hunk 24]
```diff
@@ -0,0 +1,108 @@
+""" This module allows to resolve output SEs for Job based
+on se and site/country association
```
Comment (andresailer) [inline src/DIRAC/DataManagementSystem/Utilities/ResolveSE.py]:
```suggestion
on SE and site/country association
```

File: src/DIRAC/DataManagementSystem/Utilities/ResolveSE.py [hunk 25]
```diff
@@ -0,0 +1,108 @@
...
+    if result["OK"]:
+        sLog.info("Found concrete SE", str(outputSE))
+        return [outputSE]
+
+    # Get local SEs
+    localSEs = getSEsForSite(site)
+    if not localSEs["OK"]:
+        raise RuntimeError(localSEs["Message"])
+    localSEs = localSEs["Value"]
+    sLog.verbose("Local SE list is: ", ", ".join(localSEs))
```
Comment (andresailer) [inline src/DIRAC/DataManagementSystem/Utilities/ResolveSE.py]:
```suggestion
    sLog.verbose("Local SE list is:", ", ".join(localSEs))
```

File: src/DIRAC/DataManagementSystem/Utilities/ResolveSE.py [hunk 26]
```diff
@@ -0,0 +1,108 @@
...
+def _setLocalFirst(seList, localSEs):
+    """return a shuffled list of SEs from seList, localSEs being first."""
+    # Make a copy not to change the original order
+    seList = list(seList)
+    shuffle(seList)
+    # localSEs are put first in the list
+    return sorted(seList, key=lambda x: x not in localSEs)
+
+
+def getDestinationSEList(outputSE, site, outputmode="Any"):
```
Comment (andresailer) [inline src/DIRAC/DataManagementSystem/Utilities/ResolveSE.py:22]: Document parameters and return (type)
The iLCDirac implementation still has S_OK/S_ERROR returns

File: dirac.cfg [hunk 27]
```diff
@@ -852,6 +852,162 @@ Resources
...
+
+            # If the CE is a cream CE the mode of submission
+            SubmissionMode = Direct
+
+            # Number of available processors on worker nodes
+            NumberOfProcessors = 12
+
+            # Number of available worker nodes per allocation.
+            # Values can be a number (e.g. 2 nodes) or a range of values
+            # (e.g. from 2 to 4 nodes) which lets the choice to the batch
```
Comment (andresailer) [inline dirac.cfg]:
```suggestion
            # (e.g. from 2 to 4 nodes) which leaves the choice to the batch
```

File: dirac.cfg [hunk 28]
```diff
@@ -852,6 +852,162 @@ Resources
...
+            # List of tags specific for the CE
+            Tag = GPU, 96RAM
+
+            # List of required tags that a job to be eligible must have
+            RequiredTag = GPU,96RAM
+
+            # Queues available for this VO in the CE
+            Queues
+            {
+              # Name of the queue exactly how is published
```
Comment (andresailer) [inline dirac.cfg]: Actually this is very rarely how it is published, mostly how we (I) massage it in the bdii2csagent

```suggestion
              # Name of the queue exactly how it is published
```

File: docs/source/AdministratorGuide/Resources/site.rst [hunk 29]
```diff
@@ -0,0 +1,45 @@
+.. _cs-site:
+
+Sites
+=====
+
+
+Site Names
+----------
+
+Sites have names resulting from concatenation of:
```
Comment (andresailer) [inline docs/source/AdministratorGuide/Resources/site.rst]:
```suggestion
Sites have names resulting from the concatenation of:
```

File: docs/source/AdministratorGuide/Resources/site.rst (hunk 16 above)
Comment (andresailer) [reply docs/source/AdministratorGuide/Resources/site.rst]: not necessarily an acronym

File: docs/source/AdministratorGuide/Resources/site.rst [hunk 30]
```diff
@@ -0,0 +1,45 @@
...
+* VAC.Manchester.uk
+* DIRAC.farm.cern
+
+The [Domain] may imply a (set of) technologies used for exploiting the resources, even though this is not necessarily true.
+The use of these Domains is mostly for reporting purposes,
+and it is the responsibility of the administrator of the DIRAC installation to chose them
+in such a way that they are meaningful for the communities and for the computing resources served by the installation.
+In any case, DIRAC will always be a default Domain if nothing else is specified for a given resource.
+
+The Domain, Site and the country must be unique alphanumeric strings, irrespective of case, with a possible use of the following characters: "_".
```
Comment (andresailer) [inline docs/source/AdministratorGuide/Resources/site.rst]:
```suggestion
The Domain, Site and the country must be unique alphanumeric strings, irrespective of case, with a possible use of the following characters: "_" "-".
```

File: dirac.cfg [hunk 31]
```diff
@@ -852,6 +852,162 @@ Resources
...
+        SE = CERN-RAW, CERN-RDST, CERN-USER # (Optional) SEs Local to the site
+
+        # Overwrite definities of StorageElement (discouraged)
+        # or StorageElementGroups for that Site
+        AssociatedSEs
+        {
+          # Tier1-Failover is now only CERN-FAILOVER when running a Job at CERN
+          Tier1-Failover = CERN-FAILOVER
+        }
+        CE = ce503.cern.ch, ce504.cern.ch # List of CEs for th esite (updated by BDII2CSAgent)
```
Comment (andresailer) [inline dirac.cfg]:
```suggestion
```
This list is no longer needed, cf. the CERN site entry in lbcertifdirac70 (don't look at the others, people just added them unnecessarily).
But I can't find the PR that changed this...

Comment (fstagni) [reply dirac.cfg]: This was for RSS, and the PR that removed the need for this entry was made long time ago.

**Comment by fstagni** [comment]
> @fstagni regarding all the typo and mistakes in the WMS docs, I Wanted to ask you to have a look, because I am pretty sure that a lot of things are not valid anymore. I propose that you do all the relevant changes directly on this PR, you should have push permissions

done in https://github.com/chaen/DIRAC/pull/21

File: docs/source/AdministratorGuide/Resources/site.rst (hunk 20 above)
Comment (chaen) [reply docs/source/AdministratorGuide/Resources/site.rst:14]: yes, although `cern` is the exception, and in a lot of places in the code, we use `country` so I am a bit reluctant to change it just in the doc.

**Comment by chaen** [comment]
I finally took all the review into account

**Pull request merged**